Sources/Apollo/RecordSet.swift

```diff
@@ -1,70 +1,98 @@
+import Foundation
+
+public struct RecordRow {
+  public internal(set) var record: Record
+  public internal(set) var modifiedAt: Date
+
+  public init(record: Record, modifiedAt: Date) {
+    self.record = record
+    self.modifiedAt = modifiedAt
+  }
+}
+
 /// A set of cache records.
 public struct RecordSet {
-  public private(set) var storage: [CacheKey: Record] = [:]
-  
-  public init<S: Sequence>(records: S) where S.Iterator.Element == Record {
-    insert(contentsOf: records)
+  public private(set) var storage: [CacheKey: RecordRow] = [:]
+
+  public init<S: Sequence>(rows: S) where S.Iterator.Element == RecordRow {
+    insert(contentsOf: rows)
   }
-  
+
+  init(_ dictionary: Dictionary<CacheKey, (Record.Fields, Date)>) {
+    self.init(rows: dictionary.map { RecordRow(record: Record(key: $0.0, $0.1.0), modifiedAt: $0.1.1) })
+  }
+
+  public mutating func insert(_ row: RecordRow) {
+    storage[row.record.key] = row
+  }
+
   public mutating func insert(_ record: Record) {
-    storage[record.key] = record
+    insert(.init(record: record, modifiedAt: Date()))
   }
 
   public mutating func clear() {
     storage.removeAll()
   }
-  
+
+  public mutating func insert<S: Sequence>(contentsOf rows: S) where S.Iterator.Element == RecordRow {
+    for row in rows {
+      insert(row)
+    }
+  }
+
   public mutating func insert<S: Sequence>(contentsOf records: S) where S.Iterator.Element == Record {
+    let now = Date()
     for record in records {
-      insert(record)
+      insert(RecordRow(record: record, modifiedAt: now))
     }
   }
-  
-  public subscript(key: CacheKey) -> Record? {
+
+  public subscript(key: CacheKey) -> RecordRow? {
     return storage[key]
   }
-  
+
   public var isEmpty: Bool {
     return storage.isEmpty
   }
 
   public var keys: [CacheKey] {
     return Array(storage.keys)
   }
-  
+
   @discardableResult public mutating func merge(records: RecordSet) -> Set<CacheKey> {
     var changedKeys: Set<CacheKey> = Set()
-    
-    for (_, record) in records.storage {
-      changedKeys.formUnion(merge(record: record))
+
+    for (_, row) in records.storage {
+      changedKeys.formUnion(merge(record: row.record))
     }
-    
+
     return changedKeys
   }
-  
+
   @discardableResult public mutating func merge(record: Record) -> Set<CacheKey> {
-    if var oldRecord = storage.removeValue(forKey: record.key) {
+    if var oldRow = storage.removeValue(forKey: record.key) {
       var changedKeys: Set<CacheKey> = Set()
-      
+
       for (key, value) in record.fields {
-        if let oldValue = oldRecord.fields[key], equals(oldValue, value) {
+        if let oldValue = oldRow.record.fields[key], equals(oldValue, value) {
           continue
```

Comment:
@designatednerd this part is not quite accurate as if the new record matches the existing one, we want to still update the `modifiedAt` to now to signal that this value is up to date. What is the best way to update the new value and all its references?
This also applies to `SQLiteNormalizedCache`

Reply:
Also, correct me if I'm wrong, I think that when using SQLite, this returns all the keys the first time it is called for a specific result as it calls into the in memory `RecordSet` that would be empty at this point: https://github.com/apollographql/apollo-ios/blob/master/Sources/ApolloSQLite/SQLiteNormalizedCache.swift#L54

So it runs a db update for each key, which is a trade off against running a db read and preloading the existing entry in memory, but maybe not intentional.

This  makes the changes in this PR a bit harder to do as `changedFieldKeys` is not exactly what one would think.